Fix slowness problem on the Electron version

rafaellehmkuhl opened the pull request:
This patch performs several changes in order to deal with the performance issue caused by the loading of video files that are in the Cockpit video folder:
- Stop loading all videos automatically to the memory
- Stop mounting the video-library on startup
- Loads videos to memory only when they are selected

I've also removed the download-video-button from the electron version, since the video is already on disk. This also helps directing the user to the folder button, which opens the folder that contains the videos.

After this patch, when the user clicks the video, loading it to memory usually takes less than a second (tested with 0-800MB videos). The memory usage increases during that second but goes away right after.

Electron startup before (top is memory usage in mega bytes, bottom is application frames per second):
<img src="https://github.com/user-attachments/assets/6303b02b-710e-4883-a4d4-d4c02c11f127" width="600">


After:
<img width="600" alt="image" src="https://github.com/user-attachments/assets/cc20266b-968f-4c2b-91af-4bf8da92c373" />

Startup and usage after the patch:

https://github.com/user-attachments/assets/39f46789-ca63-4afe-a401-e784a881f927

Fix #1608

ES-Alexander commented:
Nice! 😃

Minor concerns:
1. Could loading a video fully into memory when clicked on cause issues for videos that are larger than the available RAM supports?
    - IIRC part of the benefit of Cockpit's application variant over its browser variant is not being limited by memory capacity for doing long/high-res video recordings, which is kind of lost if you then can't view the files later (but I'm not sure how hard it would be to support loading/unloading chunks, and/or streaming the desired frames from the file)
2. Is there some way to have thumbnails without loading the full video into memory?
    - At the moment it looks like it would be hard to differentiate the videos without loading each of them, which seems cumbersome
    - Maybe the first frame of a recording can be saved at the start of the recording process, and deleted when the video is?
    - If/when we support #594 we could store a frame per event and have hover thumbnails for pre-scrubbing when viewing a video as well, and possibly switch the thumbnail to just the frame from the first event (when there is one), but those are future considerations

rafaellehmkuhl commented:
> Nice! 😃
> 
> Minor concerns:
> 
> 1. Could loading a video fully into memory when clicked on cause issues for videos that are larger than the available RAM supports?
>    
>    * IIRC part of the benefit of Cockpit's application variant over its browser variant is not being limited by memory capacity for doing long/high-res video recordings, which is kind of lost if you then can't view the files later (but I'm not sure how hard it would be to support loading/unloading chunks, and/or streaming the desired frames from the file)
> 2. Is there some way to have thumbnails without loading the full video into memory?
>    
>    * At the moment it looks like it would be hard to differentiate the videos without loading each of them, which seems cumbersome
>    * Maybe the first frame of a recording can be saved at the start of the recording process, and deleted when the video is?
>    * If/when we support [Eventing #594](https://github.com/bluerobotics/cockpit/issues/594) we could store a frame per event and have hover thumbnails for pre-scrubbing when viewing a video as well, and possibly switch the thumbnail to just the frame from the first event (when there is one), but those are future considerations


1. Yes. If you try to load a video bigger than your available memory, your OS will start using swap (or the Windows equivalent) and slow down, but this is expected. At some point we should implement streaming of the video file, but this will take more time, specially because this is not supported on the browser without secure context, so we will need to have separated behaviors for browser vs electron. For the time this patch solves the bug and doesn't affect the UX at all, so it should be enough.
2. Thumbnails should be working as they were before. They don't appear on my video because they are external big videos that I injected in my instance just for testing, so I didn't had thumbnails for them. The current recording pipeline already stores the first frame as the thumnail in a Base64 format. Are the thumbnails not working there?

ArturoManzoli reviewed (changes requested):
While testing the modifications on the video recording/storing/playing methods, I had the following results:

-----------------------------------

**On Electron:
TEST 1: (Videos recorder on the first Cockpit startup after starting the electron app)**
**Auto processing:**
video 1 - No thumb, not playable
video 2 - No thumb, playble
video 3 - No thumb, not playable
**Manual processing:**
video 4 - No thumb, playable
video 5 - No thumb, playable 
video 6 - No thumb, not playable
video 7 - No thumb, playable 

--------------------------------

**On Electron:
TEST 2: (Rebooted Cockpit keeping the electron app running)**
**Auto processing:**
video 1 - No thumb, playable
video 2 - No thumb, playable
video 3 - No thumb, playable
**Manual processing:** (after processing videos are not playable right away)
video 4 - thumb while unprocessed, playable only after selecting other video and then this one again
video 5 - thumb while unprocessed, playable only after selecting other video and then this one again
video 6 - thumb while unprocessed, playable only after selecting other video and then this one again

---------------------------------

**TEST 3: (Cockpit running as web app)
Same results as TEST 2**

------------------------------------------------

Main problems are the videos not having thumbs after processed and not being playable right after being processed, needing to be desselected and resselected to be playable.

Also Electron performance on first boot wasn't good for video recording. Might be just a coincidence, but most videos weren't playable.

Below: Video not playable after processing:
![image](https://github.com/user-attachments/assets/d618faeb-3ffb-464f-9f03-199c75fa80cd)

Below: Videos not playable after recording on fist Electron app startup:
![image](https://github.com/user-attachments/assets/6e093aba-c182-4649-8463-53bc3d5eeb75)

rafaellehmkuhl commented:
> While testing the modifications on the video recording/storing/playing methods, I had the following results:
> 
> **On Electron: TEST 1: (Videos recorder on the first Cockpit startup after starting the electron app)** **Auto processing:** video 1 - No thumb, not playable video 2 - No thumb, playble video 3 - No thumb, not playable **Manual processing:** video 4 - No thumb, playable video 5 - No thumb, playable video 6 - No thumb, not playable video 7 - No thumb, playable
> 
> **On Electron: TEST 2: (Rebooted Cockpit keeping the electron app running)** **Auto processing:** video 1 - No thumb, playable video 2 - No thumb, playable video 3 - No thumb, playable **Manual processing:** (after processing videos are not playable right away) video 4 - thumb while unprocessed, playable only after selecting other video and then this one again video 5 - thumb while unprocessed, playable only after selecting other video and then this one again video 6 - thumb while unprocessed, playable only after selecting other video and then this one again
> 
> **TEST 3: (Cockpit running as web app) Same results as TEST 2**
> 
> Main problems are the videos not having thumbs after processed and not being playable right after being processed, needing to be de-selected and re-selected to be playable.
> 
> Also Electron performance on first boot wasn't good for video recording. Might be just a coincidence, but half the videos weren't playable.
> 
> Below: Video not playable after processing:
> 
> Below: Videos not playable after recording on fist Electron app startup:

Will check about on the thumbs problem. 

About the performance not being good for video recording, can you check memory-usage and fps plots and compare with the same numbers on master? There shouldn't be anything worsening the performance in any case, only the opposite.

Also about the thumbs and videos being playable, can you check against master as well?

ArturoManzoli commented:
> About the performance not being good for video recording, can you check memory-usage and fps plots and compare with the same numbers on master? There shouldn't be anything worsening the performance in any case, only the opposite.
> 
> Also about the thumbs and videos being playable, can you check against master as well?

Sure, I'll check those

ArturoManzoli commented:
**Test 1 - On Master - Electron
3 videos auto processed, 3 manually.**
All videos normal but video 4, that had no thumbnail before being manually processed

https://github.com/user-attachments/assets/7a3b7dfb-6c72-4575-9785-28fca74a8017

--------------------------------------
Test 2 - On Master - Web app
3 videos auto processed, 3 manually.**
All videos normal but video 6, that had no thumbnail before being manually processed

https://github.com/user-attachments/assets/96d364c0-4f71-4d71-8706-0f09b34b2ec8

--------------------------------------
--------------------------------------
**Retesting On Branch - fix-electron-slow**
3 videos auto processed, 3 manually.**

**Test 1 - Electron App - Using app on first load**
No thumbnails generated
Video 3, 5 and 6 were not playable

https://github.com/user-attachments/assets/12b55b68-1e3b-4ecb-8b52-3f115997ada7

-------------------------------------------
**Test 2 - Electron App after reloading
No thumbnails
Only videos 2 and 6 were playable

https://github.com/user-attachments/assets/7b5da249-eeee-4e32-b442-87ac5714936b

------------------------------------------
Test 3 - Web App
Thumbnail only on video 5

Only video 6 wasn't playable

https://github.com/user-attachments/assets/5ab820f7-c70d-42ee-81b9-024d338ccce4

rafaellehmkuhl commented:
> **Test 1 - On Master - Electron 3 videos auto processed, 3 manually.** All videos normal but video 4, that had no thumbnail before being manually processed
> 
> Test 2 - On Master - Web app 3 videos auto processed, 3 manually.** All videos normal but video 6, that had no thumbnail before being manually processed
> 
> **Retesting On Branch - fix-electron-slow** 3 videos auto processed, 3 manually.**
> 
> **Test 1 - Electron App - Using app on first load** No thumbnails generated Video 3, 5 and 6 were not playable
> 
> **Test 2 - Electron App after reloading No thumbnails Only videos 2 and 6 were playable
> 
> Test 3 - Web App Thumbnail only on video 5
> 
> Only video 6 wasn't playable
> 

There were indeed some caveats into the pipeline! There was no thumbnail stored for processed videos.
Fixed that and the other problems.
Also fixed a bug that was already on master of not being able to discard unprocessed videos.
